drivers/ethernet/phy_gecko.c

```diff
@@ -0,0 +1,287 @@
+
+		k_sleep(50);
+
+		retval = phy_read(phy, MII_BMCR, &phy_reg);
+		if (retval < 0) {
+			return retval;
+		}
+	} while (phy_reg & MII_BMCR_RESET);
+
+	return 0;
+}
+
+int phy_gecko_init(const struct phy_gecko_dev *phy)
+{
+	ETH_TypeDef *const eth = phy->regs;
+	int phy_id;
+
+	mdio_bus_enable(eth);
+
+	LOG_INF("Soft Reset of ETH PHY");
+	phy_soft_reset(phy);
+
+	/* Verify that the PHY device is responding */
+	phy_id = phy_gecko_id_get(phy);
+	if (phy_id == 0xFFFFFFFF) {
+		LOG_ERR("Unable to detect a valid PHY");
+		return -1;
+	}
+
+	LOG_INF("PHYID: 0x%X at addr: %d", phy_id, phy->address);
+
+	mdio_bus_disable(eth);
+
+	return 0;
+}
+
+u32_t phy_gecko_id_get(const struct phy_gecko_dev *phy)
+{
+	ETH_TypeDef *const eth = phy->regs;
+	u32_t phy_reg;
+	u32_t phy_id;
+
+	mdio_bus_enable(eth);
+
+	if (phy_read(phy, MII_PHYID1R, &phy_reg) < 0) {
+		return 0xFFFFFFFF;
+	}
+
+	phy_id = (phy_reg & 0xFFFF) << 16;
+
+	if (phy_read(phy, MII_PHYID2R, &phy_reg) < 0) {
+		return 0xFFFFFFFF;
+	}
+
+	phy_id |= (phy_reg & 0xFFFF);
+
+	mdio_bus_disable(eth);
+
+	return phy_id;
+}
+
+int phy_gecko_auto_negotiate(const struct phy_gecko_dev *phy,
+				u32_t *status)
+{
+	ETH_TypeDef *const eth = phy->regs;
+	u32_t val;
+	u32_t ability_adv;
+	u32_t ability_rcvd;
+	u32_t retries = PHY_AUTONEG_TIMEOUT_MS / 100;
+	int retval;
+
+	mdio_bus_enable(eth);
+
+	LOG_DBG("Starting ETH PHY auto-negotiate sequence");
+
+	/* Read PHY default advertising parameters */
+	retval = phy_read(phy, MII_ANAR, &ability_adv);
+	if (retval < 0) {
+		goto auto_negotiate_exit;
+	}
+
+	/* Configure and start auto-negotiation process */
+	retval = phy_read(phy, MII_BMCR, &val);
+	if (retval < 0) {
+		goto auto_negotiate_exit;
+	}
+
+	val |= MII_BMCR_AUTONEG_ENABLE | MII_BMCR_AUTONEG_RESTART;
+	val &= ~MII_BMCR_ISOLATE;  /* Don't isolate the PHY */
+
+	retval = phy_write(phy, MII_BMCR, val);
+	if (retval < 0) {
+		goto auto_negotiate_exit;
+	}
+
+	/* Wait for the auto-negotiation process to complete */
+	do {
+		if (retries-- == 0U) {
+			retval = -ETIMEDOUT;
+			goto auto_negotiate_exit;
+		}
+
+		k_sleep(100);
+
+		retval = phy_read(phy, MII_BMSR, &val);
+		if (retval < 0) {
+			goto auto_negotiate_exit;
+		}
+	} while (!(val & MII_BMSR_AUTONEG_COMPLETE));
+
+	LOG_DBG("PHY auto-negotiate sequence completed");
+
+	/* Read abilities of the remote device */
+	retval = phy_read(phy, MII_ANLPAR, &ability_rcvd);
+	if (retval < 0) {
+		goto auto_negotiate_exit;
+	}
+
+	/* Determine the best possible mode of operation */
+	if ((ability_adv & ability_rcvd) & MII_ADVERTISE_100_FULL) {
+		*status = ETH_NETWORKCFG_FULLDUPLEX | ETH_NETWORKCFG_SPEED;
+	} else if ((ability_adv & ability_rcvd) & MII_ADVERTISE_100_HALF) {
+		*status = ETH_NETWORKCFG_SPEED;
+	} else if ((ability_adv & ability_rcvd) & MII_ADVERTISE_10_FULL) {
+		*status = ETH_NETWORKCFG_FULLDUPLEX;
+	} else {
+		*status = 0;
+	}
+
+	LOG_INF("common abilities: speed %s Mb, %s duplex",
+		*status & ETH_NETWORKCFG_FULLDUPLEX ? "full" : "half",
```

Comment:
The prints seem to be wrong order, speed should be first followed by duplex setting.

Reply:
Oops, corrected